docs/design/build-system.md

```diff
@@ -0,0 +1,162 @@
+If multiple asynchronous chunks contain a reference to the same module, it is placed in a new asynchronously loaded chunk named after the other chunks that use it.
+
+There is also a special chunk called `runtime` that contains the module loading logic and is loaded before the others.
+
+
+## Optimizations
+
+The build system contains optimizations aimed at improving the performance (for development builds) or the size of artifacts (for production builds).
+These are often mutually exclusive and thus we cannot just default to always using them.
+
+
+### Development optimizations
+
+Development optimizations focus on reducing rebuild time on watched builds.
+Although faster is always better, our threshold is to keep rebuilds even for large projects below 2 seconds.
+
+Computation needed to bundle code grows with its total size because of the cost of string concatenation and source map operations.
+Third party dependencies that are initially loaded are split into a synchronously loaded chunk called `vendor`.
+Splitting the infrequently changed vendor code from the frequently changed source code thus helps make rebuilds faster.
+
+When processing stylesheets, Webpack stores the intermediate modules as JavaScript code.
+The JavaScript wrapper code makes stylesheets larger and `mini-css-extract-plugin` must be used to obtain the actually stylesheet content into a CSS file.
+In development however, we skip the CSS extraction and leave it as JavaScript code for faster rebuild times.
+
+Watched builds split the processing load of TypeScript compilation between file emission on the main process and type checking on a forked process.
+Large projects can also opt-out of AOT compilation for faster rebuilds.
+
+
+### Production optimizations
+
+Angular CLI focuses on enabling tree-shaking (removing unused modules) and dead code elimination (removing unused module code).
+These two categories have high potential for size reduction because of network effects: removing code can lead to more code being removed.
+
+The main tool we use to achieve this goal are the compression capabilities of [Terser](https://github.com/terser/terser).
+We also use Terser's mangling, by which names, but not properties, are renamed to shorter forms.
+The main characteristics of Terser to keep in mind is that it operates via static analysis and does not support the indirection introduced by module loading.
+Thus the rest of the pipeline is directed towards providing Terser with code that can be removed via static analysis in large single modules scopes.
+
+To this end we developed [@angular-devkit/build-optimizer](https://github.com/angular/angular-cli/tree/master/packages/angular_devkit/build_optimizer), a post-processing tool for TS code.
+Build Optimizer searches for code patterns produced by the TypeScript and Angular compiler that are known to inhibit dead code elimination, and converts them into equivalent structures that enable it instead.
```

Comment:
An example of such a transformation assuming the reader knows nothing would be very useful.